Source/Layout/ASYogaUtilities.mm

```diff
@@ -0,0 +1,126 @@
+    case ASStackLayoutJustifyContentCenter:       return YGJustifyCenter;
+    case ASStackLayoutJustifyContentEnd:          return YGJustifyFlexEnd;
+    case ASStackLayoutJustifyContentSpaceBetween: return YGJustifySpaceBetween;
+    case ASStackLayoutJustifyContentSpaceAround:  return YGJustifySpaceAround;
+  }
+}
+
+YGAlign yogaAlignSelf(ASStackLayoutAlignSelf alignSelf)
+{
+  switch (alignSelf) {
+    case ASStackLayoutAlignSelfStart:   return YGAlignFlexStart;
+    case ASStackLayoutAlignSelfCenter:  return YGAlignCenter;
+    case ASStackLayoutAlignSelfEnd:     return YGAlignFlexEnd;
+    case ASStackLayoutAlignSelfStretch: return YGAlignStretch;
+    case ASStackLayoutAlignSelfAuto:    return YGAlignAuto;
+  }
+}
+
+YGFlexDirection yogaFlexDirection(ASStackLayoutDirection direction)
+{
+  return direction == ASStackLayoutDirectionVertical ? YGFlexDirectionColumn : YGFlexDirectionRow;
+}
+
+float yogaFloatForCGFloat(CGFloat value)
+{
+  if (value < CGFLOAT_MAX / 2) {
+    return value;
+  } else {
+    return YGUndefined;
+  }
+}
+
+float yogaDimensionToPoints(ASDimension dimension)
+{
+  ASDisplayNodeCAssert(dimension.unit == ASDimensionUnitPoints,
+                       @"Dimensions should not be type Fraction for this method: %f", dimension.value);
+  return yogaFloatForCGFloat(dimension.value);
+}
+
+float yogaDimensionToPercent(ASDimension dimension)
+{
+  ASDisplayNodeCAssert(dimension.unit == ASDimensionUnitFraction,
+                       @"Dimensions should not be type Points for this method: %f", dimension.value);
+  return 100.0 * yogaFloatForCGFloat(dimension.value);
+
+}
+
+ASDimension dimensionForEdgeWithEdgeInsets(YGEdge edge, ASEdgeInsets insets)
+{
+  switch (edge) {
+    case YGEdgeLeft:          return insets.left;
+    case YGEdgeTop:           return insets.top;
+    case YGEdgeRight:         return insets.right;
+    case YGEdgeBottom:        return insets.bottom;
+    case YGEdgeStart:         return insets.start;
+    case YGEdgeEnd:           return insets.end;
+    case YGEdgeHorizontal:    return insets.horizontal;
+    case YGEdgeVertical:      return insets.vertical;
+    case YGEdgeAll:           return insets.all;
+    default: ASDisplayNodeCAssert(NO, @"YGEdge other than ASEdgeInsets is not supported.");
+      return ASDimensionAuto;
+  }
+}
+
+YGSize ASLayoutElementYogaMeasureFunc(YGNodeRef yogaNode, float width, YGMeasureMode widthMode,
+                                      float height, YGMeasureMode heightMode)
+{
+  id <ASLayoutElement> layoutElement = (__bridge id <ASLayoutElement>)YGNodeGetContext(yogaNode);
+  ASSizeRange sizeRange;
+  sizeRange.min = CGSizeZero;
+  sizeRange.max = CGSizeMake(width, height);
+  if (widthMode == YGMeasureModeExactly) {
+    sizeRange.min.width = sizeRange.max.width;
+  }
+  if (heightMode == YGMeasureModeExactly) {
+    sizeRange.min.height = sizeRange.max.height;
+  }
+  ASDisplayNodeCAssert([layoutElement isKindOfClass:[ASDisplayNode class]], @"!");
```

Comment:
Thinking again, I believe we should do `sizeRange.min.(width|height) = layoutElement.style.minSize.(width|height)` when `(width|height)Mode` is either `YGMeasureModeAtMost` or `YGMeasureModeUndefined`.

The reason is that, even if width and height were resolved against their min values, in the case of these modes, a sub-element still needs a min constraint to measure against. And `style.minSize` will not be picked-up by the sub-element if it's not included in `rootConstrainedSize`.